From 6fe2715ff3630ead239ed2bacbe1e831bfd6303d Mon Sep 17 00:00:00 2001
From: Rafael Palomar <rafael.palomar@ous-research.no>
Date: Mon, 6 Feb 2023 20:21:04 +0100
Subject: [PATCH 23/23] ENH: Update Slicer build macros

This (1) adds libraries and (2) sets link directories for Slicer build
macros.
---
 CMake/SlicerMacroBuildApplication.cmake    | 3 +++
 CMake/SlicerMacroBuildLoadableModule.cmake | 4 ++++
 CMake/SlicerMacroBuildModuleLogic.cmake    | 8 ++++++++
 CMake/SlicerMacroBuildModuleWidgets.cmake  | 7 +++++++
 4 files changed, 22 insertions(+)

diff --git a/CMake/SlicerMacroBuildApplication.cmake b/CMake/SlicerMacroBuildApplication.cmake
index 7ee7bc11ca..f79b7c8066 100644
--- a/CMake/SlicerMacroBuildApplication.cmake
+++ b/CMake/SlicerMacroBuildApplication.cmake
@@ -442,6 +442,9 @@ macro(slicerMacroBuildApplication)
     LABELS ${SLICERAPP_NAME}
     OUTPUT_NAME ${executable_name}
     )
+  set_target_properties(${slicerapp_target} PROPERTIES
+    INSTALL_RPATH
+    "${Slicer_INSTALL_QTLOADABLEMODULES_LIB_DIR}")
 
   if(APPLE)
     set(link_flags "-Wl,-rpath,@loader_path/../")
diff --git a/CMake/SlicerMacroBuildLoadableModule.cmake b/CMake/SlicerMacroBuildLoadableModule.cmake
index 624955260c..b73b117c05 100644
--- a/CMake/SlicerMacroBuildLoadableModule.cmake
+++ b/CMake/SlicerMacroBuildLoadableModule.cmake
@@ -188,6 +188,9 @@ macro(slicerMacroBuildLoadableModule)
   # --------------------------------------------------------------------------
   # Build library
   # --------------------------------------------------------------------------
+
+  link_directories(${Slicer_INSTALL_QTLOADABLEMODULES_LIB_DIR})
+
   add_library(${lib_name}
     ${LOADABLEMODULE_SRCS}
     ${LOADABLEMODULE_MOC_OUTPUT}
@@ -203,6 +206,7 @@ macro(slicerMacroBuildLoadableModule)
     ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${Slicer_QTLOADABLEMODULES_LIB_DIR}"
     )
   set_target_properties(${lib_name} PROPERTIES LABELS ${lib_name})
+  set_target_properties(${lib_name} PROPERTIES INSTALL_RPATH "${Slicer_INSTALL_QTLOADABLEMODULES_LIB_DIR}")
 
   target_link_libraries(${lib_name}
     # The two PUBLIC keywords are not a duplication, they allow developers to
diff --git a/CMake/SlicerMacroBuildModuleLogic.cmake b/CMake/SlicerMacroBuildModuleLogic.cmake
index 56b443918d..08e5b21f05 100644
--- a/CMake/SlicerMacroBuildModuleLogic.cmake
+++ b/CMake/SlicerMacroBuildModuleLogic.cmake
@@ -53,6 +53,7 @@ macro(SlicerMacroBuildModuleLogic)
     ${Slicer_Base_INCLUDE_DIRS}
     ${Slicer_ModuleLogic_INCLUDE_DIRS}
     ${Slicer_ModuleMRML_INCLUDE_DIRS}
+    ${vtkAddon_INCLUDE_DIRS}
     )
 
   if(Slicer_BUILD_CLI_SUPPORT)
@@ -69,6 +70,10 @@ macro(SlicerMacroBuildModuleLogic)
 
     list(APPEND MODULELOGIC_TARGET_LIBRARIES
       qSlicerBaseQTCLI
+      ${VTK_LIBRARIES}
+      ${vtkAddon_LIBRARIES}
+      SlicerBaseLogic
+      MRMLLogic
       )
     # HACK Explicitly list transitive VTK dependencies because _get_dependencies_recurse
     # used in vtkAddon/CMake/vtkMacroKitPythonWrap.cmake only recurses over dependencies
@@ -95,6 +100,8 @@ macro(SlicerMacroBuildModuleLogic)
     set(MODULELOGIC_NO_INSTALL_OPTION "NO_INSTALL")
   endif()
 
+  link_directories(${Slicer_INSTALL_QTLOADABLEMODULES_LIB_DIR})
+
   SlicerMacroBuildModuleVTKLibrary(
     NAME ${MODULELOGIC_NAME}
     EXPORT_DIRECTIVE ${MODULELOGIC_EXPORT_DIRECTIVE}
@@ -105,6 +112,7 @@ macro(SlicerMacroBuildModuleLogic)
     ${MODULELOGIC_NO_INSTALL_OPTION}
     )
 
+  set_target_properties(${MODULELOGIC_NAME} PROPERTIES INSTALL_RPATH "${Slicer_INSTALL_QTLOADABLEMODULES_LIB_DIR}")
   set_property(GLOBAL APPEND PROPERTY SLICER_MODULE_LOGIC_TARGETS ${MODULELOGIC_NAME})
 
   #-----------------------------------------------------------------------------
diff --git a/CMake/SlicerMacroBuildModuleWidgets.cmake b/CMake/SlicerMacroBuildModuleWidgets.cmake
index 506fdb71e8..73d62ef330 100644
--- a/CMake/SlicerMacroBuildModuleWidgets.cmake
+++ b/CMake/SlicerMacroBuildModuleWidgets.cmake
@@ -57,10 +57,13 @@ macro(SlicerMacroBuildModuleWidgets)
     ${Slicer_ModuleLogic_INCLUDE_DIRS}
     ${Slicer_ModuleMRML_INCLUDE_DIRS}
     ${Slicer_ModuleWidgets_INCLUDE_DIRS}
+    ${vtkAddon_INCLUDE_DIRS}
     )
 
   list(APPEND MODULEWIDGETS_TARGET_LIBRARIES
     ${Slicer_GUI_LIBRARY}
+    ${vtkAddon_LIBRARIES}
+    Qt::Xml
     )
 
   if(NOT DEFINED MODULEWIDGETS_FOLDER AND DEFINED MODULE_NAME)
@@ -101,6 +104,8 @@ macro(SlicerMacroBuildModuleWidgets)
     set(QM_OUTPUT_FILES )
   endif()
 
+  link_directories(${Slicer_INSTALL_QTLOADABLEMODULES_LIB_DIR})
+
   # --------------------------------------------------------------------------
   # Build library
   # --------------------------------------------------------------------------
@@ -119,6 +124,8 @@ macro(SlicerMacroBuildModuleWidgets)
     )
 
   set_property(GLOBAL APPEND PROPERTY SLICER_MODULE_WIDGET_TARGETS ${MODULEWIDGETS_NAME})
+  set_target_properties(${MODULEWIDGETS_NAME} PROPERTIES INSTALL_RPATH "${Slicer_INSTALL_QTLOADABLEMODULES_LIB_DIR}")
+
 
   #-----------------------------------------------------------------------------
   # Update Slicer_ModuleWidgets_INCLUDE_DIRS
-- 
2.39.1

